### PR TITLE
Foundry Start-to-End Tutorial

### PR DESCRIPTION
**Pull request by jboetticher**
### Description

Added a new tutorial that takes a slightly more complex look at Foundry.

### Checklist

- [x] If this requires translations for the `moonbeam-docs-cn` repo, I have created a ticket for the translations in Jira
- [x] If images have been added, I have run the `compress-images.py` script to compress the images.

### Corresponding PRs

Please link to any corresponding PRs here.

### After Translation Requirements

- [ ] Will need to create PR in `moonbeam-docs` repo to remove images
- [ ] Will need to create PR in `moonbeam-docs` repo to remove variables
- [ ] Will need to create PR in `moonbeam-mkdocs` repo to add redirects for Chinese site
- [ ] No additional PRs are required after the translations are done

#### Items to be Updated

Please list any of the items that will need to be added or deleted after the translations are done here.

### CODE REVIEW

File: tutorials/foundry-start-to-end.md [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+### Unit Tests in Foundry {: #unit-tests-in-foundry}
+
+To get started with writing tests for this tutorial, make a new file in the `test` folder:  
+
+```
+cd test
+touch MyToken.t.sol
+```
+
+By convention, all of your test should end with `.t.sol` and start with the name of the smart contract that it is testing. In practice, the test can be stored anywhere, and is considered a test if it has a function that starts with the word *"test"*. Let's start by writing a test for the token smart contract. Open up `MyToken.t.sol` and add:
```
Comment (themacexpert) [inline tutorials/foundry-start-to-end.md:174]:
```suggestion
By convention, all of your tests should end with `.t.sol` and start with the name of the smart contract that it is testing. In practice, the test can be stored anywhere, and is considered a test if it has a function that starts with the word *"test"*. Let's start by writing a test for the token smart contract. Open up `MyToken.t.sol` and add:  
```

File: tutorials/foundry-start-to-end.md [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+        token = new MyToken(100);
+    }
+
+    function testConstructorMint() public {
+        assertEq(token.balanceOf(address(this)), 100);
+    }
+}
+```
+
+Let's break down what's happening here. The first line is typical for a Solidity file: setting the Solidity version. The next two lines are imports. `forge-std/Test.sol` is the standard library that forge (and thus Foundry) includes to help with testing. This includes the `Test` smart contract, certain assertations, and [forge cheatcodes](https://book.getfoundry.sh/forge/cheatcodes).
```
Comment (themacexpert) [inline tutorials/foundry-start-to-end.md:195]:
```suggestion
Let's break down what's happening here. The first line is typical for a Solidity file: setting the Solidity version. The next two lines are imports. `forge-std/Test.sol` is the standard library that forge (and thus Foundry) includes to help with testing. This includes the `Test` smart contract, certain assertions, and [forge cheatcodes](https://book.getfoundry.sh/forge/cheatcodes).  
```

File: tutorials/foundry-start-to-end.md [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+
+    function testConstructorMint() public {
+        assertEq(token.balanceOf(address(this)), 100);
+    }
+}
+```
+
+Let's break down what's happening here. The first line is typical for a Solidity file: setting the Solidity version. The next two lines are imports. `forge-std/Test.sol` is the standard library that forge (and thus Foundry) includes to help with testing. This includes the `Test` smart contract, certain assertations, and [forge cheatcodes](https://book.getfoundry.sh/forge/cheatcodes).  
+
+If you take a look at the `MyTokenTest` smart contract, you'll see two functions. The first is `setUp`, which is ran before each test. So in this test contract, a new instance of `MyToken` is deployed every time a test function is ran. You know if a function is a test function if it starts with the word *"test"*, so the second function, `testConstructorMint` is a test function.
```
Comment (themacexpert) [inline tutorials/foundry-start-to-end.md:197]:
```suggestion
If you take a look at the `MyTokenTest` smart contract, you'll see two functions. The first is `setUp`, which is run before each test. So in this test contract, a new instance of `MyToken` is deployed every time a test function is run. You know if a function is a test function if it starts with the word *"test"*, so the second function, `testConstructorMint` is a test function.  
```

File: tutorials/foundry-start-to-end.md [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+
+**PUT THE CONSOLE IMAGE HERE**  
+
+### Fuzzing Tests in Foundry {: #fuzzing-tests-in-foundry}
+
+### Forking Tests in Foundry {: #forking-tests-in-foundry}
+
+## Deploying in Foundry with Solidity Scripts {: #deploying-in-foundry-with-solidity-scripts }
+
+Not only are tests in Foundry written in Solidity, the scripts are too! Like other developer environments, scripts can be written to help interact with deployed smart contracts or can help along a complex deployment process that would be difficult to do manually. Even though scripts are written in Solidity, they are never deployed to a chain. Instead, much of the logic is actually ran off-chain, so don't worry about any additional gas costs for using Foundry instead of a JavaScript environment like HardHat.
```
Comment (themacexpert) [inline tutorials/foundry-start-to-end.md:288]:
```suggestion
Not only are tests in Foundry written in Solidity, the scripts are too! Like other developer environments, scripts can be written to help interact with deployed smart contracts or can help along a complex deployment process that would be difficult to do manually. Even though scripts are written in Solidity, they are never deployed to a chain. Instead, much of the logic is actually run off-chain, so don't worry about any additional gas costs for using Foundry instead of a JavaScript environment like HardHat.  
```

File: tutorials/foundry-start-to-end.md [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+        new Container(token, 500);
+
+        vm.stopBroadcast();
+    }
+}
+```
+
+Let's break this script down. The first line is once again standard: declaring the solidity version. The imports include the two smart contracts you previously added, which will be deployed. This includes additional functionality to use in a Script, including the `Script` contract.  
+
+Now let's look at the logic in the contract. There is a single function, `run`, which is where the script logic is hosted. In this `run` function, the `vm` object is used often. This is where all of the Forge cheatcodes are stored, which determines the state of the virtual machine that the solidity is ran in.
```
Comment (themacexpert) [inline tutorials/foundry-start-to-end.md:327]:
```suggestion
Now let's look at the logic in the contract. There is a single function, `run`, which is where the script logic is hosted. In this `run` function, the `vm` object is used often. This is where all of the Forge cheatcodes are stored, which determines the state of the virtual machine that the solidity is run in.  
```

File: tutorials/foundry-start-to-end.md [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+
+# Using Foundry Start to End with Moonbeam
+
+![Banner Image](/images/tutorials/foundry-start-to-end/foundry-banner.png)
+
+_January 10th, 2022 | by Jeremy Boetticher_
+
+## Introduction {: #introduction } 
+
+Foundry has become an increasingly popular developer environment to develop smart contracts with, since utilizing it only requires a single language: Solidity. Moonbeam offers [introductory documentation on using Foundry](/builders/build/eth-api/dev-env/foundry.md) with Moonbeam networks, which is recommended to read to get an introduction to using Foundry. In this tutorial we will be dipping our toes deeper into the library to get a more cohesive look at how to properly develop, test, and deploy.
```
Comment (themacexpert) [inline tutorials/foundry-start-to-end.md:14]:
```suggestion
Foundry has become an increasingly popular developer environment to develop smart contracts with, since utilizing it only requires a single language: Solidity. Moonbeam offers [introductory documentation on using Foundry](/builders/build/eth-api/dev-env/foundry) with Moonbeam networks, which is recommended to read to get an introduction to using Foundry. In this tutorial we will be dipping our toes deeper into the library to get a more cohesive look at how to properly develop, test, and deploy.  
```

File: tutorials/foundry-start-to-end.md [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+
+[etherscan]
+moonbase = { key = "${MOONSCAN_API_KEY}" }
+```
+
+The first addition is a specification of the `solc_version`, underneath `profile.default`. The `rpc_endpoints` tag allows you to define which RPC endpoints to use when deploying to a named network, in this case, Moonbase Alpha. The `etherscan` tag allows you to add Etherscan API keys for smart contract verification, which we will go over later.  
+
+## Writing Smart Contracts {: #writing-smart-contracts-in-foundry }
+
+Smart contracts in Foundry that are meant to be deployed will by default belong in the `src` folder. In this tutorial, we'll write two smart contracts. Starting with the token:
```
Comment (themacexpert) [inline tutorials/foundry-start-to-end.md:79]:
```suggestion
Smart contracts in Foundry that are meant to be deployed by default belong in the `src` folder. In this tutorial, we'll write two smart contracts. Starting with the token:
```

File: tutorials/foundry-start-to-end.md [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+}
+```
+
+The `Container` smart contract can have its status updated based on how many tokens it holds. It requires a `MyToken` smart contract instance to function, so when we deploy it, we will need logic to ensure that it is deployed with a `MyToken` smart contract.  
+
+## Testing in Foundry {: #testing-in-foundry }
+
+Before we deploy anything to a TestNet or MainNet, however, it's good to test your smart contracts. There are many types of tests:
+
+- Unit tests are tests that allow you to test specific parts of a smart contract's functionality. When writing your own smart contracts, it can be a good idea to break functionality into different sections so that it is easier to unit test
```
Comment (themacexpert) [inline tutorials/foundry-start-to-end.md:161]: Perhaps we could bold or italicize the names of the tests? Check with @eshaben ...

Comment (themacexpert) [inline tutorials/foundry-start-to-end.md:161]:
```suggestion
- Unit tests allow you to test specific parts of a smart contract's functionality. When writing your own smart contracts, it can be a good idea to break functionality into different sections so that it is easier to unit test
```

File: tutorials/foundry-start-to-end.md [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+- Fuzz tests are tests that allow you to test a smart contract with a wide variety of inputs to check for edge cases
```
Comment (themacexpert) [inline tutorials/foundry-start-to-end.md:162]:
```suggestion
- Fuzz tests allow you to test a smart contract with a wide variety of inputs to check for edge cases
```

File: tutorials/foundry-start-to-end.md [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+- Integration tests are tests that allow you to test a smart contract when it works in conjunction with other smart contracts, so that you know it works as expected in a deployed environment. Often you can do this by forking an EVM, so that you can simulate a transaction on a MainNet
```
Comment (themacexpert) [inline tutorials/foundry-start-to-end.md:163]:
```suggestion
- Integration tests allow you to test a smart contract when it works in conjunction with other smart contracts, so that you know it works as expected in a deployed environment. Often you can do this by forking an EVM, so that you can simulate a transaction on a MainNet
```

File: tutorials/foundry-start-to-end.md [hunk 11]
```diff
@@ -0,0 +1,365 @@
...
+
+        // Make a new container
+        new Container(token, 500);
+
+        vm.stopBroadcast();
+    }
+}
+```
+
+Let's break this script down. The first line is once again standard: declaring the solidity version. The imports include the two smart contracts you previously added, which will be deployed. This includes additional functionality to use in a Script, including the `Script` contract.
```
Comment (themacexpert) [inline tutorials/foundry-start-to-end.md:325]:
```suggestion
Let's break this script down. The first line is standard: declaring the solidity version. The imports include the two smart contracts you previously added, which will be deployed. This includes additional functionality to use in a Script, including the `Script` contract.  
```

File: tutorials/foundry-start-to-end.md [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+
+And within this file, add the following:  
+
+```
+PRIVATE_KEY=YOUR_PRIVATE_KEY
+MOONSCAN_API_KEY=YOUR_MOONSCAN_API_KEY
+```
+
+!!! note
+    Foundry provides [options for handling your private key](https://book.getfoundry.sh/reference/forge/forge-script#wallet-options---raw), other than providing it in the environment variables. It is up to you to decide whether or not you would rather use it in the console, have it stored in your device's environment, using a hardware wallet, or using a keystore.
```
Comment (themacexpert) [inline tutorials/foundry-start-to-end.md:345]:
```suggestion
    Foundry provides [additional options for handling your private key](https://book.getfoundry.sh/reference/forge/forge-script#wallet-options---raw). It is up to you to decide whether or not you would rather use it in the console, have it stored in your device's environment, using a hardware wallet, or using a keystore.
```

**Review by themacexpert — CHANGES_REQUESTED** [review]

File: .snippets/code/tutorials/foundry-start-to-end/Container.t.sol [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+        // Get token that's already deployed & deploys a container instance
+        token = MyToken(0x93e1e9EC6c1A8736266A595EFe97B5673ea0fEAc);
+        container = new Container(token, CAPACITY);
+
+        // Mint tokens to the container & update container status
+        token.mint(CAPACITY, address(container));
+        container.updateStatus();
+
+        // Assert that the capacity is full, just like the rest of the time.
```
Comment (eshaben) [inline .snippets/code/tutorials/foundry-start-to-end/Container.t.sol:53]:
```suggestion
        // Assert that the capacity is full, just like the rest of the time
```

File: builders/build/eth-api/dev-env/foundry.md [hunk 14]
```diff
@@ -274,4 +274,6 @@ From here you can deploy new contracts to your forked instance of Moonbeam or in
 cast call INSERT-CONTRACT-ADDRESS  "balanceOf(address)(uint256)" INSERT-YOUR-ADDRESS --rpc-url http://localhost:8545
 ```
 
+If you want to learn more about Foundry, try following the [Foundry start-to-end tutorial](/tutorials/foundry-start-to-end) in our tutorial section.
```
Comment (eshaben) [inline builders/build/eth-api/dev-env/foundry.md:277]:
```suggestion
If you want to learn more about Foundry, try following the [Foundry start-to-end tutorial](/tutorials/foundry-start-to-end){target=_blank} in our tutorial section. 
```

File: tutorials/foundry-start-to-end.md [hunk 15]
```diff
@@ -0,0 +1,458 @@
...
+// Import OpenZeppelin Contract
+import {ERC20} from "openzeppelin-contracts/contracts/token/ERC20/ERC20.sol";
+
+// This ERC-20 contract mints the specified amount of tokens to the contract creator
+contract MyToken is ERC20 {
+  constructor(uint256 initialSupply) ERC20("MyToken", "MYTOK") {
+    _mint(msg.sender, initialSupply);
+  }
+
+  // An external minting function allows anyone to mint as many tokens as they want.
```
Comment (eshaben) [inline tutorials/foundry-start-to-end.md:100]:
```suggestion
  // An external minting function allows anyone to mint as many tokens as they want
```

File: tutorials/foundry-start-to-end.md [hunk 16]
```diff
@@ -0,0 +1,458 @@
...
+
+        // Get token that's already deployed & deploys a container instance
+        token = MyToken(0x359436610E917e477D73d8946C2A2505765ACe90);
+        container = new Container(token, CAPACITY);
+
+        // Mint tokens to the container & update container status
+        token.mint(CAPACITY, address(container));
+        container.updateStatus();
+
+        // Assert that the capacity is full, just like the rest of the time.
```
Comment (eshaben) [inline tutorials/foundry-start-to-end.md:355]:
```suggestion
        // Assert that the capacity is full, just like the rest of the time
```

File: tutorials/foundry-start-to-end.md [hunk 17]
```diff
@@ -0,0 +1,458 @@
...
+The first step (and thus first line) in this function is to have the test function fork a network with `vm.createFork`. Recall that `vm` is a cheatcode provided by the Forge standard library. All that's necessary to create a fork is an RPC URL, or an alias for an RPC URL that's stored in the `foundry.toml` file. In this case, we added an RPC URL for "moonbase" in [the setup step](#setup-a-foundry-project), so in the test function we will just pass the word "moonbase". This cheatcode function returns an ID for the fork created, which is necessary for activating the fork.  
+
+On the second line, after the fork has been created, the environment will select and use the fork in the test environment with `vm.selectFork`. The third line is just to demonstrate that the current fork, retrieved with `vm.activeFork`, is the same as the Moonbase Alpha fork.  
+
+The fourth line of code retrieves an already deployed instance of `MyToken`, which is what's useful about forking: you can use contracts that are already deployed.  
+
+The rest of the code tests capacity like you would expect a local test to. If you run the tests (with the "-vvvv" tag for extra logging), you'll see that it passes:
```
Comment (eshaben) [inline tutorials/foundry-start-to-end.md:367]:
```suggestion
The rest of the code tests capacity like you would expect a local test to. If you run the tests (with the `-vvvv` tag for extra logging), you'll see that it passes:  
```

File: tutorials/foundry-start-to-end.md [hunk 18]
```diff
@@ -0,0 +1,458 @@
...
+        token.mint(CAPACITY, address(container));
+        container.updateStatus();
+
+        // Assert that the capacity is full, just like the rest of the time.
+        assertEq(token.balanceOf(address(container)), CAPACITY);
+        assertTrue(container.status() == ContainerStatus.Full);
+    }
+```
+
+The first step (and thus first line) in this function is to have the test function fork a network with `vm.createFork`. Recall that `vm` is a cheatcode provided by the Forge standard library. All that's necessary to create a fork is an RPC URL, or an alias for an RPC URL that's stored in the `foundry.toml` file. In this case, we added an RPC URL for "moonbase" in [the setup step](#setup-a-foundry-project), so in the test function we will just pass the word "moonbase". This cheatcode function returns an ID for the fork created, which is necessary for activating the fork.
```
Comment (eshaben) [inline tutorials/foundry-start-to-end.md:361]:
```suggestion
The first step (and thus first line) in this function is to have the test function fork a network with `vm.createFork`. Recall that `vm` is a cheatcode provided by the Forge standard library. All that's necessary to create a fork is an RPC URL, or an alias for an RPC URL that's stored in the `foundry.toml` file. In this case, we added an RPC URL for "moonbase" in [the setup step](#setup-a-foundry-project), so in the test function we will just pass the word `"moonbase"`. This cheatcode function returns an ID for the fork created, which is necessary for activating the fork.  
```

File: tutorials/foundry-start-to-end.md [hunk 19]
```diff
@@ -0,0 +1,458 @@
...
+
+        // Make a new container
+        new Container(token, 500);
+
+        vm.stopBroadcast();
+    }
+}
+```
+
+Let's break this script down. The first line is standard: declaring the solidity version. The imports include the two smart contracts you previously added, which will be deployed. This includes additional functionality to use in a Script, including the `Script` contract.
```
Comment (eshaben) [inline tutorials/foundry-start-to-end.md:416]:
```suggestion
Let's break this script down. The first line is standard: declaring the solidity version. The imports include the two smart contracts you previously added, which will be deployed. This includes additional functionality to use in a script, including the `Script` contract.  
```

File: tutorials/foundry-start-to-end.md [hunk 20]
```diff
@@ -0,0 +1,458 @@
...
+source .env
+```
+
+Now your script and project should be ready for deployment! Use the following command to do so:  
+
+```
+forge script Container.s.sol:ContainerDeployScript --broadcast --verify -vvvv --rpc-url moonbase
+```
+
+What this command does is run the `ContainerDeployScript` contract as a script. The `--broadcast` option tells Forge to allow broadcasting of transactions, the `--verify` option tells Forge to verify to Moonscan when deploying, `-vvvv` makes the command output verbose, and `--rpc-url moonbase` sets the network to what *"moonbase"* was set to in `foundry.toml`.
```
Comment (eshaben) [inline tutorials/foundry-start-to-end.md:450]:
```suggestion
What this command does is run the `ContainerDeployScript` contract as a script. The `--broadcast` option tells Forge to allow broadcasting of transactions, the `--verify` option tells Forge to verify to Moonscan when deploying, `-vvvv` makes the command output verbose, and `--rpc-url moonbase` sets the network to what `moonbase` was set to in `foundry.toml`.  
```

**Review by eshaben — CHANGES_REQUESTED** [review]
Ahh I had already reviewed this but never submitted my review. 🤦‍♀️ So, this includes my previous review plus I took time to review the new sections you added. 

Please double check throughout the guide and add `{target=_blank}` to any missing links

File: tutorials/foundry-start-to-end.md [hunk 21]
```diff
@@ -0,0 +1,458 @@
...
+// Import OpenZeppelin Contract
+import {MyToken} from "./MyToken.sol";
+
+enum ContainerStatus {
+    Unsatisfied,
+    Full,
+    Overflowing
+}
+
+// This ERC-20 contract mints the specified amount of tokens to the contract creator
```
Comment (jboetticher) [inline tutorials/foundry-start-to-end.md:131]:
```suggestion
```

File: tutorials/foundry-start-to-end.md [hunk 22]
```diff
@@ -0,0 +1,365 @@
...
+To add these environment variables, run the following command:  
+
+```
+source .env
+```
+
+Now your script and project should be ready for deployment! Use the following command to do so:  
+
+```
+forge script script/Container.s.sol:ContainerDeployScript --broadcast --verify -vvvv --rpc-url moonbase
```
[the remark this answers is not in the document]
Comment (jboetticher) [reply tutorials/foundry-start-to-end.md]: I was able to do the whole thing with a fresh virtual box, so this might be a bug with Foundry? Did you try running after updating foundry with `foundryup`?

[the remark this answers is not in the document]
Comment (eshaben) [reply tutorials/foundry-start-to-end.md]: awesome that worked! thank you!

File: tutorials/foundry-start-to-end.md [hunk 23]
```diff
@@ -0,0 +1,470 @@
...
+
+# Using Foundry Start to End with Moonbeam
+
+![Banner Image](/images/tutorials/foundry-start-to-end/foundry-banner.png)
+
+_January 10th, 2022 | by Jeremy Boetticher_
+
+## Introduction {: #introduction } 
+
+Foundry has become an increasingly popular developer environment to develop smart contracts with, since utilizing it only requires a single language: Solidity. Moonbeam offers [introductory documentation on using Foundry](/builders/build/eth-api/dev-env/foundry) with Moonbeam networks, which is recommended to read to get an introduction to using Foundry. In this tutorial we will be dipping our toes deeper into the library to get a more cohesive look at how to properly develop, test, and deploy.
```
Comment (eshaben) [inline tutorials/foundry-start-to-end.md:14]:
```suggestion
Foundry has become an increasingly popular developer environment to develop smart contracts with, since utilizing it only requires a single language: Solidity. Moonbeam offers [introductory documentation on using Foundry](/builders/build/eth-api/dev-env/foundry){target=_blank} with Moonbeam networks, which is recommended to read to get an introduction to using Foundry. In this tutorial we will be dipping our toes deeper into the library to get a more cohesive look at how to properly develop, test, and deploy.  
```

File: tutorials/foundry-start-to-end.md [hunk 24]
```diff
@@ -0,0 +1,470 @@
...
+
+The `Container` smart contract can have its status updated based on how many tokens it holds. It requires a `MyToken` smart contract instance to function, so when we deploy it, we will need logic to ensure that it is deployed with a `MyToken` smart contract.  
+
+## Testing in Foundry {: #testing-in-foundry }
+
+Before we deploy anything to a TestNet or MainNet, however, it's good to test your smart contracts. There are many types of tests:
+
+- **Unit tests** — allow you to test specific parts of a smart contract's functionality. When writing your own smart contracts, it can be a good idea to break functionality into different sections so that it is easier to unit test
+- **Fuzz tests** — allow you to test a smart contract with a wide variety of inputs to check for edge cases
+- **Integration tests** — allow you to test a smart contract when it works in conjunction with other smart contracts, so that you know it works as expected in a deployed environment.
```
Comment (eshaben) [inline tutorials/foundry-start-to-end.md:170]:
```suggestion
- **Integration tests** — allow you to test a smart contract when it works in conjunction with other smart contracts, so that you know it works as expected in a deployed environment
```

**Review by eshaben — APPROVED** [review]
Woo! This looks good!

My only thing is that the `foundry-5.png` image is a bit small and I know it's probably because the output is really long, but can you possibly do some editing magic to make it so the text a bit bigger?

Besides that LGTM ✨

**Review by albertov19 — CHANGES_REQUESTED** [review]
Looks great! Love the detailed section on testing! Left some feedback

File: tutorials/foundry-start-to-end.md [hunk 25]
```diff
@@ -0,0 +1,470 @@
...
+        assertTrue(container.status() == ContainerStatus.Full);
+    }
+}
+```
+
+This test smart contract has two tests, so when running the tests, there will be two deployments of both `MyToken` and `Container`, for four smart contracts in total. When testing, you should see the following command:  
+
+![Unit Testing in Foundry](/images/tutorials/foundry-start-to-end/foundry-1.png)
+
+### Test Harnesses in Foundry {: #test-harnesses-in-foundry }
```
[the remark this answers is not in the document]
Comment (jboetticher) [reply tutorials/foundry-start-to-end.md:266]: It's more of a subset of a Unit test, but I can make it more clear that it is one

File: tutorials/foundry-start-to-end.md [hunk 26]
```diff
@@ -0,0 +1,470 @@
...
+
+## Setup a Foundry Project {: #setup-a-foundry-project }
+
+The first step to start a Foundry project is of course to create it. If you have Foundry installed, you can run:
+
+```
+forge init foundry && cd foundry
+```
+
+This will have the forge utility initialize a new folder named `foundry` with a Foundry project initialized within it. The `script`, `src`, and `test` folders may have files in them already. Be sure to delete them, because we will be writing our own soon.
```
[the remark this answers is not in the document]
Comment (eshaben) [reply tutorials/foundry-start-to-end.md:37]: Yeah, I think it makes sense to put it in back ticks here because were referring to the `forge` command

File: tutorials/foundry-start-to-end.md [hunk 27]
```diff
@@ -0,0 +1,488 @@
...
+moonbeam = "https://rpc.api.moobeam.network"
+
+[etherscan]
+moonbase = { key = "${MOONSCAN_API_KEY}" }
+moonbeam = { key = "${MOONSCAN_API_KEY}" }
+```
+
+The first addition is a specification of the `solc_version`, underneath `profile.default`. The `rpc_endpoints` tag allows you to define which RPC endpoints to use when deploying to a named network, in this case, Moonbase Alpha and Moonbeam. The `etherscan` tag allows you to add Etherscan API keys for smart contract verification, which we will go over later.  
+
+## Adding Smart Contracts {: #adding-smart-contracts-in-foundry }
```
Comment (eshaben) [inline tutorials/foundry-start-to-end.md:79]:
```suggestion
## Add Smart Contracts {: #add-smart-contracts-in-foundry }
```

File: tutorials/foundry-start-to-end.md [hunk 28]
```diff
@@ -0,0 +1,488 @@
...
+
+```
+forge test --vvvv
+```
+
+![Forking Tests in Foundry](/images/tutorials/foundry-start-to-end/foundry-4.png)
+
+That's it for testing! You can see the complete [`Container.t.sol` file](https://raw.githubusercontent.com/PureStake/moonbeam-docs/master/.snippets/code/tutorials/foundry-start-to-end/Container.t.sol){target=_blank} and [`MyToken.t.sol` file](https://raw.githubusercontent.com/PureStake/moonbeam-docs/master/.snippets/code/tutorials/foundry-start-to-end/MyToken.t.sol){target=_blank} on GitHub.
+
+## Deploying in Foundry with Solidity Scripts {: #deploying-in-foundry-with-solidity-scripts }
```
Comment (eshaben) [inline tutorials/foundry-start-to-end.md:394]:
```suggestion
## Deploy in Foundry with Solidity Scripts {: #deploy-in-foundry-with-solidity-scripts }
```

Comment (eshaben) [reply tutorials/foundry-start-to-end.md:394]: I'll update the "deploying" titles in my PR to "deploy"

File: tutorials/foundry-start-to-end.md [hunk 29]
```diff
@@ -0,0 +1,488 @@
...
+
+You should see something like this as output:  
+
+![Running a Script in Foundry](/images/tutorials/foundry-start-to-end/foundry-5.png)
+
+You should be able to see that your contracts were deployed, and are verified on Moonscan! For example, this is where my [`Container.sol` contract was deployed](https://moonbase.moonscan.io/address/0xe8bf2e654d7c1c1ba8f55fed280ddd241e46ced9#code){target=_blank}.  
+
+The entire deployment script is [available on GitHub](https://raw.githubusercontent.com/PureStake/moonbeam-docs/master/.snippets/code/tutorials/foundry-start-to-end/Container.s.sol){target=_blank}.  
+
+### Deploying on Moonbeam MainNet {: #deploying-on-moonbeam-mainnet }
```
Comment (eshaben) [inline tutorials/foundry-start-to-end.md:478]:
```suggestion
### Deploy on Moonbeam MainNet {: #deploy-on-moonbeam-mainnet }
```

File: tutorials/foundry-start-to-end.md [hunk 30]
```diff
@@ -0,0 +1,488 @@
...
+
+Let's say that you're comfortable with your smart contracts and you want to deploy on the Moonbeam MainNet! The process isn't too different from what was just done, you just have to change the command's rpc-url:
```
Comment (eshaben) [inline tutorials/foundry-start-to-end.md:480]: I would also make a note that they'll need to update their configs from `"moonbase"` to `"moonbeam"`

Comment (eshaben) [reply tutorials/foundry-start-to-end.md:480]: even though it's obvious 😅

File: tutorials/foundry-start-to-end.md [hunk 31]
```diff
@@ -0,0 +1,488 @@
...
+
+## Deploying in Foundry with Solidity Scripts {: #deploying-in-foundry-with-solidity-scripts }
+
+Not only are tests in Foundry written in Solidity, the scripts are too! Like other developer environments, scripts can be written to help interact with deployed smart contracts or can help along a complex deployment process that would be difficult to do manually. Even though scripts are written in Solidity, they are never deployed to a chain. Instead, much of the logic is actually run off-chain, so don't worry about any additional gas costs for using Foundry instead of a JavaScript environment like HardHat.  
+
+In this tutorial, we will be using Foundry's scripts to deploy both the `MyToken` and `Container` smart contracts. To create the deployment scripts, create a new file in the `script` folder:
```
Comment (eshaben) [inline tutorials/foundry-start-to-end.md:398]: maybe add a subtitle before this line for Moonbase Alpha

**Review by eshaben — CHANGES_REQUESTED** [review]
Almost there! image 5 looks nice! haha ✨💜

File: tutorials/foundry-start-to-end.md [hunk 32]
```diff
@@ -0,0 +1,490 @@
+---
+title: Using Foundry Start to End with Moonbeam
```
Comment (eshaben) [inline tutorials/foundry-start-to-end.md:2]:
```suggestion
title: Using Foundry Start to End
```

**Review by eshaben — APPROVED** [review]
LGTM ✨ 🚀 🚢
